[Site Isolation] Escape doesn't exit fullscreen with site isolated fullscreen content

nmoucht opened the pull request:
#### c9c6fd9ba28a57dc52e30138802cf87fd282f3ce
<pre>
[Site Isolation] Escape doesn&apos;t exit fullscreen with site isolated fullscreen content
<a href="https://bugs.webkit.org/show_bug.cgi?id=293989">https://bugs.webkit.org/show_bug.cgi?id=293989</a>
<a href="https://rdar.apple.com/150307508">rdar://150307508</a>

Reviewed by Tim Nguyen.

Hitting escape routes the key event to the correct process and eventually calls down into
DocumentFullscreen::fullyExitFullscreen, which currently doesn&apos;t work for subframe processes,
as mainFrameDocument() will return null. Instead use the root frame&apos;s document for this case.

* LayoutTests/http/tests/site-isolation/fullscreen-escape-key-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/fullscreen-escape-key.html: Added.
* LayoutTests/http/tests/site-isolation/resources/fullscreen-escape-key.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::fullyExitFullscreen):

Canonical link: <a href="https://commits.webkit.org/296004@main">https://commits.webkit.org/296004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/434b0c0e6c8451d109dc0425339f1e22896ed442

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111864 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108694 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34908 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80990 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109659 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21474 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96234 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61326 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20927 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14341 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56707 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90814 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114739 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33793 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24916 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90055 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34157 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92465 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89763 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34681 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29420 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17303 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33718 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39131 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33464 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36817 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/DocumentFullscreen.cpp:
```diff
@@ -612,7 +612,7 @@ void DocumentFullscreen::fullyExitFullscreen()
     RefPtr mainFrameDocument = this->mainFrameDocument();
     if (!mainFrameDocument)
-        LOG_ONCE(SiteIsolation, "Unable to fully perform DocumentFullscreen::fullyExitFullscreen() without access to the main frame document ");
+        mainFrameDocument = document().frame()->rootFrame().document();
```
We should null check document.frame()
If we do this, we should rename the variable name to rootFrameDocument.
If we do this, we don't need to try to get the mainFrameDocument any more.

nmoucht replied:
Fixed.

nt1m reviewed (commented):
The code change looks good. Fullscreening from subframes propagates to the root frame, so this is the correct thing to do for this case.

Is there a way we can test this?

If this comes from the web process (e.g. EventHandler), then you can probably just simulate pressing the Escape key like `fullscreen/requestFullscreen-escape-key.html`.

If this comes from the UI process, then you should be able to follow something among those lines: https://github.com/WebKit/WebKit/commit/daad2275a5d51c2c4bde7776c695e3b02bc3e623

nt1m commented on LayoutTests/http/tests/site-isolation/fullscreen-escape-key.html:
```diff
@@ -0,0 +1,21 @@
+<script>
+    addEventListener("message", (event) => {                
+        if (event.data.startsWith('Entered') && window.testRunner) { eventSender.keyDown("escape"); }
+        if (event.data.startsWith('Exited') && window.testRunner) { testRunner.notifyDone(); }
```
```suggestion
        if (event.data.startsWith("Entered") && window.testRunner) { eventSender.keyDown("escape"); }
        if (event.data.startsWith("Exited") && window.testRunner) { testRunner.notifyDone(); }
```

nmoucht replied:
Fixed.

nt1m commented on LayoutTests/http/tests/site-isolation/fullscreen-escape-key.html:
```diff
@@ -0,0 +1,21 @@
+    }
+
+</script>
+<iframe onload="test()" id='frame' allowfullscreen src="http://localhost:8000/site-isolation/resources/fullscreen-escape-key.html" frameborder=0></iframe>
```
```suggestion
<iframe onload="test()" id="frame" allowfullscreen src="http://localhost:8000/site-isolation/resources/fullscreen-escape-key.html" frameborder=0></iframe>
```

nt1m replied:
I wonder if you could also just use `./resources/fullscreen-escape-key.html` as src

nmoucht replied:
This doesn't expose the bug unfortunately.

nt1m commented on LayoutTests/http/tests/site-isolation/resources/fullscreen-escape-key.html:
```diff
@@ -0,0 +1,16 @@
+        if (window.testRunner) { await testRunner.updatePresentation() }
+    }
+
+    document.documentElement.addEventListener("fullscreenchange", () => { if (!document.fullscreenElement) { window.parent.postMessage("Exited",  "*"); }});
```
```suggestion
    document.documentElement.addEventListener("fullscreenchange", () => { if (!document.fullscreenElement) { window.parent.postMessage("Exited", "*"); }});
```

nmoucht replied:
Fixed.

nmoucht committed:
[Site Isolation] Escape doesn't exit fullscreen with site isolated fullscreen content
https://bugs.webkit.org/show_bug.cgi?id=293989
rdar://150307508

Reviewed by Tim Nguyen.

Hitting escape routes the key event to the correct process and eventually calls down into
DocumentFullscreen::fullyExitFullscreen, which currently doesn't work for subframe processes,
as mainFrameDocument() will return null. Instead use the root frame's document for this case.

* LayoutTests/http/tests/site-isolation/fullscreen-escape-key-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/fullscreen-escape-key.html: Added.
* LayoutTests/http/tests/site-isolation/resources/fullscreen-escape-key.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::fullyExitFullscreen):

Canonical link: https://commits.webkit.org/296004@main
